File: src/agents/auth-profiles/usage.ts
import type { OpenClawConfig } from "../../config/config.js";
import { normalizeProviderId } from "../model-selection.js";
import { saveAuthProfileStore, updateAuthProfileStoreWithLock } from "./store.js";
import type { AuthProfileFailureReason, AuthProfileStore, ProfileUsageStats } from "./types.js";

export function resolveProfileUnusableUntil(
  stats: Pick<ProfileUsageStats, "cooldownUntil" | "disabledUntil">,
): number | null {
  const values = [stats.cooldownUntil, stats.disabledUntil]
    .filter((value): value is number => typeof value === "number")
    .filter((value) => Number.isFinite(value) && value > 0);
  if (values.length === 0) {
    return null;
  }
  return Math.max(...values);
}

/**
 * Check if a profile is currently in cooldown (due to rate limiting or errors).
 */
export function isProfileInCooldown(store: AuthProfileStore, profileId: string): boolean {
  const stats = store.usageStats?.[profileId];
  if (!stats) {
    return false;
  }
  const unusableUntil = resolveProfileUnusableUntil(stats);
  return unusableUntil ? Date.now() < unusableUntil : false;
}

/**
 * Check if a profile is in cooldown for a specific model. Billing/auth
 * disables apply to all models, but rate-limit / timeout cooldowns are
 * checked per-model when a model key is provided.
 */
export function isProfileInCooldownForModel(
  store: AuthProfileStore,
  profileId: string,
  model?: string,
): boolean {
  const stats = store.usageStats?.[profileId];
  if (!stats) {
    return false;
  }

  // Billing/auth disabled applies to ALL models
  if (
    typeof stats.disabledUntil === "number" &&
    Number.isFinite(stats.disabledUntil) &&
    stats.disabledUntil > 0 &&
    Date.now() < stats.disabledUntil
  ) {
    return true;
  }

  // No model specified → fall back to global cooldown check
  if (!model) {
    return isProfileInCooldown(store, profileId);
  }

  // Check model-specific cooldown
  const mc = stats.modelCooldowns?.[model];
  if (mc?.cooldownUntil && Date.now() < mc.cooldownUntil) {
    return true;
  }

  return false;
}

/**
 * Return the soonest `unusableUntil` timestamp (ms epoch) among the given
 * profiles, or `null` when no profile has a recorded cooldown. Note: the
 * returned timestamp may be in the past if the cooldown has already expired.
 */
export function getSoonestCooldownExpiry(
  store: AuthProfileStore,
  profileIds: string[],
): number | null {
  let soonest: number | null = null;
  const consider = (until: unknown) => {
    if (typeof until !== "number" || !Number.isFinite(until) || until <= 0) {
      return;
    }
    if (soonest === null || until < soonest) {
      soonest = until;
    }
  };
  for (const id of profileIds) {
    const stats = store.usageStats?.[id];
    if (!stats) {
      continue;
    }
    consider(resolveProfileUnusableUntil(stats));
    // Also consider per-model cooldowns so probe timing accounts for them
    if (stats.modelCooldowns) {
      for (const mc of Object.values(stats.modelCooldowns)) {
        consider(mc.cooldownUntil);
      }
    }
  }
  return soonest;
}

/**
 * Clear expired cooldowns from all profiles in the store.
 *
 * When `cooldownUntil` or `disabledUntil` has passed, the corresponding fields
 * are removed and error counters are reset so the profile gets a fresh start
 * (circuit-breaker half-open → closed). Without this, a stale `errorCount`
 * causes the *next* transient failure to immediately escalate to a much longer
 * cooldown — the root cause of profiles appearing "stuck" after rate limits.
 *
 * `cooldownUntil` and `disabledUntil` are handled independently: if a profile
 * has both and only one has expired, only that field is cleared.
 *
 * Mutates the in-memory store; disk persistence happens lazily on the next
 * store write (e.g. `markAuthProfileUsed` / `markAuthProfileFailure`), which
 * matches the existing save pattern throughout the auth-profiles module.
 *
 * @returns `true` if any profile was modified.
 */
export function clearExpiredCooldowns(store: AuthProfileStore, now?: number): boolean {
  const usageStats = store.usageStats;
  if (!usageStats) {
    return false;
  }

  const ts = now ?? Date.now();
  let mutated = false;

  for (const [profileId, stats] of Object.entries(usageStats)) {
    if (!stats) {
      continue;
    }

    let profileMutated = false;
    const cooldownExpired =
      typeof stats.cooldownUntil === "number" &&
      Number.isFinite(stats.cooldownUntil) &&
      stats.cooldownUntil > 0 &&
      ts >= stats.cooldownUntil;
    const disabledExpired =
      typeof stats.disabledUntil === "number" &&
      Number.isFinite(stats.disabledUntil) &&
      stats.disabledUntil > 0 &&
      ts >= stats.disabledUntil;

    if (cooldownExpired) {
      stats.cooldownUntil = undefined;
      profileMutated = true;
    }
    if (disabledExpired) {
      stats.disabledUntil = undefined;
      stats.disabledReason = undefined;
      profileMutated = true;
    }

    // Clear expired model-level cooldowns
    if (stats.modelCooldowns) {
      for (const [modelKey, mc] of Object.entries(stats.modelCooldowns)) {
        if (
          typeof mc.cooldownUntil === "number" &&
          Number.isFinite(mc.cooldownUntil) &&
          mc.cooldownUntil > 0 &&
          ts >= mc.cooldownUntil
        ) {
          delete stats.modelCooldowns[modelKey];
          profileMutated = true;
        }
      }
      // Remove the map entirely if empty
      if (Object.keys(stats.modelCooldowns).length === 0) {
        stats.modelCooldowns = undefined;
      }
    }

    // Reset error counters when ALL cooldowns have expired so the profile gets
    // a fair retry window. Preserves lastFailureAt for the failureWindowMs
    // decay check in computeNextProfileUsageStats.
    if (profileMutated && !resolveProfileUnusableUntil(stats)) {
      stats.errorCount = 0;
      stats.failureCounts = undefined;
    }

    if (profileMutated) {
      usageStats[profileId] = stats;
      mutated = true;
    }
  }

  return mutated;
}

/**
 * Mark a profile as successfully used. Resets error count and updates lastUsed.
 * Uses store lock to avoid overwriting concurrent usage updates.
 */
export async function markAuthProfileUsed(params: {
  store: AuthProfileStore;
  profileId: string;
  agentDir?: string;
}): Promise<void> {
  const { store, profileId, agentDir } = params;
  const updated = await updateAuthProfileStoreWithLock({
    agentDir,
    updater: (freshStore) => {
      if (!freshStore.profiles[profileId]) {
        return false;
      }
      freshStore.usageStats = freshStore.usageStats ?? {};
      freshStore.usageStats[profileId] = {
        ...freshStore.usageStats[profileId],
        lastUsed: Date.now(),
        errorCount: 0,
        cooldownUntil: undefined,
        disabledUntil: undefined,
        disabledReason: undefined,
        failureCounts: undefined,
        modelCooldowns: undefined,
      };
      return true;
    },
  });
  if (updated) {
    store.usageStats = updated.usageStats;
    return;
  }
  if (!store.profiles[profileId]) {
    return;
  }

  store.usageStats = store.usageStats ?? {};
  store.usageStats[profileId] = {
    ...store.usageStats[profileId],
    lastUsed: Date.now(),
    errorCount: 0,
    cooldownUntil: undefined,
    disabledUntil: undefined,
    disabledReason: undefined,
    failureCounts: undefined,
    modelCooldowns: undefined,
  };
  saveAuthProfileStore(store, agentDir);
}

export function calculateAuthProfileCooldownMs(
  errorCount: number,
  reason?: AuthProfileFailureReason,
): number {
  const normalized = Math.max(1, errorCount);

  // Timeout uses a shorter cooldown — likely transient network/provider issue
  if (reason === "timeout") {
    return Math.min(
      5 * 60 * 1000, // 5 min max
      30 * 1000 * 2 ** Math.min(normalized - 1, 4),
    );
  }

  // Rate limit / other: gentler 2^n backoff with jitter (was 5^n)
  const base = Math.min(
    15 * 60 * 1000, // 15 min max (was 1 hour)
    60 * 1000 * 2 ** Math.min(normalized - 1, 4),
  );
  const jitter = base * (0.1 + Math.random() * 0.1);
  return Math.floor(base + jitter);
}

type ResolvedAuthCooldownConfig = {
  billingBackoffMs: number;
  billingMaxMs: number;
  failureWindowMs: number;
};

function resolveAuthCooldownConfig(params: {
  cfg?: OpenClawConfig;
  providerId: string;
}): ResolvedAuthCooldownConfig {
  const defaults = {
    billingBackoffHours: 5,
    billingMaxHours: 24,
    failureWindowHours: 24,
  } as const;

  const resolveHours = (value: unknown, fallback: number) =>
    typeof value === "number" && Number.isFinite(value) && value > 0 ? value : fallback;

  const cooldowns = params.cfg?.auth?.cooldowns;
  const billingOverride = (() => {
    const map = cooldowns?.billingBackoffHoursByProvider;
    if (!map) {
      return undefined;
    }
    for (const [key, value] of Object.entries(map)) {
      if (normalizeProviderId(key) === params.providerId) {
        return value;
      }
    }
    return undefined;
  })();

  const billingBackoffHours = resolveHours(
    billingOverride ?? cooldowns?.billingBackoffHours,
    defaults.billingBackoffHours,
  );
  const billingMaxHours = resolveHours(cooldowns?.billingMaxHours, defaults.billingMaxHours);
  const failureWindowHours = resolveHours(
    cooldowns?.failureWindowHours,
    defaults.failureWindowHours,
  );

  return {
    billingBackoffMs: billingBackoffHours * 60 * 60 * 1000,
    billingMaxMs: billingMaxHours * 60 * 60 * 1000,
    failureWindowMs: failureWindowHours * 60 * 60 * 1000,
  };
}

function calculateAuthProfileBillingDisableMsWithConfig(params: {
  errorCount: number;
  baseMs: number;
  maxMs: number;
}): number {
  const normalized = Math.max(1, params.errorCount);
  const baseMs = Math.max(60_000, params.baseMs);
  const maxMs = Math.max(baseMs, params.maxMs);
  const exponent = Math.min(normalized - 1, 10);
  const raw = baseMs * 2 ** exponent;
  return Math.min(maxMs, raw);
}

export function resolveProfileUnusableUntilForDisplay(
  store: AuthProfileStore,
  profileId: string,
): number | null {
  const stats = store.usageStats?.[profileId];
  if (!stats) {
    return null;
  }
  return resolveProfileUnusableUntil(stats);
}

function computeNextProfileUsageStats(params: {
  existing: ProfileUsageStats;
  now: number;
  reason: AuthProfileFailureReason;
  cfgResolved: ResolvedAuthCooldownConfig;
  model?: string;
  /** Server-provided Retry-After delay (ms). When available, used instead of calculated backoff. */
  retryAfterMs?: number;
}): ProfileUsageStats {
  const windowMs = params.cfgResolved.failureWindowMs;
  const windowExpired =
    typeof params.existing.lastFailureAt === "number" &&
    params.existing.lastFailureAt > 0 &&
    params.now - params.existing.lastFailureAt > windowMs;

  const baseErrorCount = windowExpired ? 0 : (params.existing.errorCount ?? 0);
  const nextErrorCount = baseErrorCount + 1;
  const failureCounts = windowExpired ? {} : { ...params.existing.failureCounts };
  failureCounts[params.reason] = (failureCounts[params.reason] ?? 0) + 1;

  const updatedStats: ProfileUsageStats = {
    ...params.existing,
    errorCount: nextErrorCount,
    failureCounts,
    lastFailureAt: params.now,
  };

  if (params.reason === "billing") {
    const billingCount = failureCounts.billing ?? 1;
    const backoffMs = calculateAuthProfileBillingDisableMsWithConfig({
      errorCount: billingCount,
      baseMs: params.cfgResolved.billingBackoffMs,
      maxMs: params.cfgResolved.billingMaxMs,
    });
    updatedStats.disabledUntil = params.now + backoffMs;
    updatedStats.disabledReason = "billing";
  } else {
    // Compute backoff: prefer server-provided Retry-After, then calculated
    let backoffMs: number;
    if (params.retryAfterMs && params.retryAfterMs > 0) {
      // Cap server-provided value at 15 min to avoid pathologically long waits
      backoffMs = Math.min(params.retryAfterMs, 15 * 60 * 1000);
    } else if (params.reason === "timeout") {
      const timeoutCount = failureCounts.timeout ?? 1;
      backoffMs = calculateAuthProfileCooldownMs(timeoutCount, "timeout");
    } else {
      backoffMs = calculateAuthProfileCooldownMs(nextErrorCount, params.reason);
    }

    if (params.model) {
      // Record per-model cooldown instead of global profile cooldown
      const existingMc = updatedStats.modelCooldowns ?? {};
      const prevEntry = existingMc[params.model] ?? {};
      existingMc[params.model] = {
        cooldownUntil: params.now + backoffMs,
        errorCount: (prevEntry.errorCount ?? 0) + 1,
        lastFailureAt: params.now,
      };
      updatedStats.modelCooldowns = existingMc;
    } else {
      updatedStats.cooldownUntil = params.now + backoffMs;
    }
  }

  return updatedStats;
}

/**
 * Mark a profile as failed for a specific reason. Billing failures are treated
 * as "disabled" (longer backoff) vs the regular cooldown window.
 */
export async function markAuthProfileFailure(params: {
  store: AuthProfileStore;
  profileId: string;
  reason: AuthProfileFailureReason;
  cfg?: OpenClawConfig;
  agentDir?: string;
  /** When set, cooldown is recorded per-model instead of globally on the profile. */
  model?: string;
  /** Server-provided Retry-After delay (ms). Overrides calculated backoff when present. */
  retryAfterMs?: number;
}): Promise<void> {
  const { store, profileId, reason, agentDir, cfg, model, retryAfterMs } = params;
  const updated = await updateAuthProfileStoreWithLock({
    agentDir,
    updater: (freshStore) => {
      const profile = freshStore.profiles[profileId];
      if (!profile) {
        return false;
      }
      freshStore.usageStats = freshStore.usageStats ?? {};
      const existing = freshStore.usageStats[profileId] ?? {};

      const now = Date.now();
      const providerKey = normalizeProviderId(profile.provider);
      const cfgResolved = resolveAuthCooldownConfig({
        cfg,
        providerId: providerKey,
      });

      freshStore.usageStats[profileId] = computeNextProfileUsageStats({
        existing,
        now,
        reason,
        cfgResolved,
        model,
        retryAfterMs,
      });
      return true;
    },
  });
  if (updated) {
    store.usageStats = updated.usageStats;
    return;
  }
  if (!store.profiles[profileId]) {
    return;
  }

  store.usageStats = store.usageStats ?? {};
  const existing = store.usageStats[profileId] ?? {};
  const now = Date.now();
  const providerKey = normalizeProviderId(store.profiles[profileId]?.provider ?? "");
  const cfgResolved = resolveAuthCooldownConfig({
    cfg,
    providerId: providerKey,
  });

  store.usageStats[profileId] = computeNextProfileUsageStats({
    existing,
    now,
    reason,
    cfgResolved,
    model,
    retryAfterMs,
  });
  saveAuthProfileStore(store, agentDir);
}

/**
 * Mark a profile as failed/rate-limited. Applies exponential backoff cooldown.
 * Cooldown times: 1min, 2min, 4min, 8min, max 15 min.
 * Uses store lock to avoid overwriting concurrent usage updates.
 */
export async function markAuthProfileCooldown(params: {
  store: AuthProfileStore;
  profileId: string;
  agentDir?: string;
}): Promise<void> {
  await markAuthProfileFailure({
    store: params.store,
    profileId: params.profileId,
    reason: "unknown",
    agentDir: params.agentDir,
  });
}

/**
 * Clear cooldown for a profile (e.g., manual reset).
 * Uses store lock to avoid overwriting concurrent usage updates.
 */
export async function clearAuthProfileCooldown(params: {
  store: AuthProfileStore;
  profileId: string;
  agentDir?: string;
}): Promise<void> {
  const { store, profileId, agentDir } = params;
  const updated = await updateAuthProfileStoreWithLock({
    agentDir,
    updater: (freshStore) => {
      if (!freshStore.usageStats?.[profileId]) {
        return false;
      }

      freshStore.usageStats[profileId] = {
        ...freshStore.usageStats[profileId],
        errorCount: 0,
        cooldownUntil: undefined,
        modelCooldowns: undefined,
      };
      return true;
    },
  });
  if (updated) {
    store.usageStats = updated.usageStats;
    return;
  }
  if (!store.usageStats?.[profileId]) {
    return;
  }

  store.usageStats[profileId] = {
    ...store.usageStats[profileId],
    errorCount: 0,
    cooldownUntil: undefined,
    modelCooldowns: undefined,
  };
  saveAuthProfileStore(store, agentDir);
}

File: openclaw_py/agents/auth_profiles/usage.py
"""Auth profile usage tracking and cooldown management."""

import time

from openclaw_py.agents.model_selection import normalize_provider_id
from openclaw_py.config import OpenClawConfig

from .store import save_auth_profile_store, update_auth_profile_store_with_lock
from .types import AuthProfileFailureReason, AuthProfileStore, ProfileUsageStats


def _resolve_profile_unusable_until(stats: ProfileUsageStats) -> int | None:
    """Resolve profile unusable until timestamp.

    Args:
        stats: ProfileUsageStats

    Returns:
        Timestamp (ms) or None
    """
    values = [stats.cooldown_until, stats.disabled_until]
    valid_values = [v for v in values if isinstance(v, (int, float)) and v > 0]
    return int(max(valid_values)) if valid_values else None


def is_profile_in_cooldown(store: AuthProfileStore, profile_id: str) -> bool:
    """Check if a profile is currently in cooldown.

    Args:
        store: AuthProfileStore
        profile_id: Profile ID

    Returns:
        True if in cooldown

    Examples:
        >>> is_profile_in_cooldown(store, "anthropic:default")
        False
    """
    if not store.usage_stats:
        return False

    stats = store.usage_stats.get(profile_id)
    if not stats:
        return False

    unusable_until = _resolve_profile_unusable_until(stats)
    if not unusable_until:
        return False

    now = int(time.time() * 1000)
    return now < unusable_until


async def mark_auth_profile_used(
    store: AuthProfileStore,
    profile_id: str,
    agent_dir: str | None = None,
) -> None:
    """Mark a profile as successfully used.

    Resets error count and updates lastUsed timestamp.

    Args:
        store: AuthProfileStore
        profile_id: Profile ID
        agent_dir: Optional agent directory
    """

    def updater(fresh_store: AuthProfileStore) -> bool:
        if profile_id not in fresh_store.profiles:
            return False

        fresh_store.usage_stats = fresh_store.usage_stats or {}
        existing = fresh_store.usage_stats.get(profile_id)

        fresh_store.usage_stats[profile_id] = ProfileUsageStats(
            last_used=int(time.time() * 1000),
            error_count=0,
            cooldown_until=None,
            disabled_until=None,
            disabled_reason=None,
            failure_counts=None,
        )
        return True

    updated = await update_auth_profile_store_with_lock(
        {"agent_dir": agent_dir, "updater": updater}
    )

    if updated:
        store.usage_stats = updated.usage_stats
        return

    # Fallback: direct update
    if profile_id not in store.profiles:
        return

    store.usage_stats = store.usage_stats or {}
    store.usage_stats[profile_id] = ProfileUsageStats(
        last_used=int(time.time() * 1000),
        error_count=0,
        cooldown_until=None,
        disabled_until=None,
        disabled_reason=None,
        failure_counts=None,
    )
    save_auth_profile_store(store, agent_dir)


def calculate_auth_profile_cooldown_ms(error_count: int) -> int:
    """Calculate cooldown duration based on error count.

    Uses exponential backoff: 1min, 5min, 25min, max 1 hour.

    Args:
        error_count: Number of consecutive errors

    Returns:
        Cooldown duration in milliseconds

    Examples:
        >>> calculate_auth_profile_cooldown_ms(1)
        60000  # 1 minute
        >>> calculate_auth_profile_cooldown_ms(2)
        300000  # 5 minutes
    """
    normalized = max(1, error_count)
    # 5^0 = 1min, 5^1 = 5min, 5^2 = 25min, 5^3 = 125min (capped at 60min)
    return min(
        60 * 60 * 1000,  # 1 hour max
        60 * 1000 * (5 ** min(normalized - 1, 3)),
    )


def _resolve_auth_cooldown_config(cfg: OpenClawConfig | None, provider_id: str) -> dict:
    """Resolve auth cooldown configuration.

    Args:
        cfg: OpenClaw configuration
        provider_id: Provider ID

    Returns:
        Dict with cooldown config values
    """
    defaults = {
        "billing_backoff_hours": 5,
        "billing_max_hours": 24,
        "failure_window_hours": 24,
    }

    if not cfg or not hasattr(cfg, "auth") or not cfg.auth:
        return {
            "billing_backoff_ms": defaults["billing_backoff_hours"] * 60 * 60 * 1000,
            "billing_max_ms": defaults["billing_max_hours"] * 60 * 60 * 1000,
            "failure_window_ms": defaults["failure_window_hours"] * 60 * 60 * 1000,
        }

    auth_config = cfg.auth
    cooldowns = auth_config.cooldowns if hasattr(auth_config, "cooldowns") else None

    # Get billing backoff hours (with provider-specific override)
    billing_backoff_hours = defaults["billing_backoff_hours"]
    if cooldowns:
        if hasattr(cooldowns, "billing_backoff_hours"):
            billing_backoff_hours = cooldowns.billing_backoff_hours or billing_backoff_hours

        # Check provider-specific override
        if hasattr(cooldowns, "billing_backoff_hours_by_provider"):
            by_provider = cooldowns.billing_backoff_hours_by_provider
            if by_provider:
                for key, value in by_provider.items():
                    if normalize_provider_id(key) == provider_id:
                        billing_backoff_hours = value or billing_backoff_hours
                        break

    billing_max_hours = defaults["billing_max_hours"]
    if cooldowns and hasattr(cooldowns, "billing_max_hours"):
        billing_max_hours = cooldowns.billing_max_hours or billing_max_hours

    failure_window_hours = defaults["failure_window_hours"]
    if cooldowns and hasattr(cooldowns, "failure_window_hours"):
        failure_window_hours = cooldowns.failure_window_hours or failure_window_hours

    return {
        "billing_backoff_ms": int(billing_backoff_hours * 60 * 60 * 1000),
        "billing_max_ms": int(billing_max_hours * 60 * 60 * 1000),
        "failure_window_ms": int(failure_window_hours * 60 * 60 * 1000),
    }


def _calculate_auth_profile_billing_disable_ms(
    error_count: int,
    base_ms: int,
    max_ms: int,
) -> int:
    """Calculate billing error disable duration.

    Uses exponential backoff: 2^0, 2^1, 2^2, ...

    Args:
        error_count: Number of billing errors
        base_ms: Base duration in ms
        max_ms: Maximum duration in ms

    Returns:
        Disable duration in milliseconds
    """
    normalized = max(1, error_count)
    base_ms = max(60_000, base_ms)  # At least 1 minute
    max_ms = max(base_ms, max_ms)
    exponent = min(normalized - 1, 10)
    raw = base_ms * (2**exponent)
    return min(max_ms, raw)


def _compute_next_profile_usage_stats(
    existing: ProfileUsageStats,
    now: int,
    reason: AuthProfileFailureReason,
    cfg_resolved: dict,
) -> ProfileUsageStats:
    """Compute next usage stats after failure.

    Args:
        existing: Existing stats
        now: Current timestamp (ms)
        reason: Failure reason
        cfg_resolved: Resolved config

    Returns:
        Updated ProfileUsageStats
    """
    window_ms = cfg_resolved["failure_window_ms"]

    # Check if failure window expired
    window_expired = (
        existing.last_failure_at
        and existing.last_failure_at > 0
        and now - existing.last_failure_at > window_ms
    )

    # Reset counts if window expired
    base_error_count = 0 if window_expired else (existing.error_count or 0)
    next_error_count = base_error_count + 1

    failure_counts = {} if window_expired else dict(existing.failure_counts or {})
    failure_counts[reason] = failure_counts.get(reason, 0) + 1

    updated_stats = ProfileUsageStats(
        last_used=existing.last_used,
        error_count=next_error_count,
        failure_counts=failure_counts,
        last_failure_at=now,
        cooldown_until=existing.cooldown_until,
        disabled_until=existing.disabled_until,
        disabled_reason=existing.disabled_reason,
    )

    # Handle billing failures specially
    if reason == "billing":
        billing_count = failure_counts.get("billing", 1)
        backoff_ms = _calculate_auth_profile_billing_disable_ms(
            billing_count,
            cfg_resolved["billing_backoff_ms"],
            cfg_resolved["billing_max_ms"],
        )
        updated_stats.disabled_until = now + backoff_ms
        updated_stats.disabled_reason = "billing"
    else:
        # Regular cooldown
        backoff_ms = calculate_auth_profile_cooldown_ms(next_error_count)
        updated_stats.cooldown_until = now + backoff_ms

    return updated_stats


async def mark_auth_profile_failure(
    store: AuthProfileStore,
    profile_id: str,
    reason: AuthProfileFailureReason,
    cfg: OpenClawConfig | None = None,
    agent_dir: str | None = None,
) -> None:
    """Mark a profile as failed for a specific reason.

    Billing failures trigger longer backoff than regular failures.

    Args:
        store: AuthProfileStore
        profile_id: Profile ID
        reason: Failure reason
        cfg: OpenClaw configuration
        agent_dir: Optional agent directory
    """

    def updater(fresh_store: AuthProfileStore) -> bool:
        profile = fresh_store.profiles.get(profile_id)
        if not profile:
            return False

        fresh_store.usage_stats = fresh_store.usage_stats or {}
        existing = fresh_store.usage_stats.get(profile_id) or ProfileUsageStats()

        now = int(time.time() * 1000)
        provider_key = normalize_provider_id(profile.provider)
        cfg_resolved = _resolve_auth_cooldown_config(cfg, provider_key)

        fresh_store.usage_stats[profile_id] = _compute_next_profile_usage_stats(
            existing, now, reason, cfg_resolved
        )
        return True

    updated = await update_auth_profile_store_with_lock(
        {"agent_dir": agent_dir, "updater": updater}
    )

    if updated:
        store.usage_stats = updated.usage_stats
        return

    # Fallback: direct update
    profile = store.profiles.get(profile_id)
    if not profile:
        return

    store.usage_stats = store.usage_stats or {}
    existing = store.usage_stats.get(profile_id) or ProfileUsageStats()

    now = int(time.time() * 1000)
    provider_key = normalize_provider_id(profile.provider)
    cfg_resolved = _resolve_auth_cooldown_config(cfg, provider_key)

    store.usage_stats[profile_id] = _compute_next_profile_usage_stats(
        existing, now, reason, cfg_resolved
    )
    save_auth_profile_store(store, agent_dir)


async def mark_auth_profile_cooldown(
    store: AuthProfileStore,
    profile_id: str,
    agent_dir: str | None = None,
) -> None:
    """Mark a profile as failed/rate-limited.

    Convenience wrapper for mark_auth_profile_failure with "unknown" reason.

    Args:
        store: AuthProfileStore
        profile_id: Profile ID
        agent_dir: Optional agent directory
    """
    await mark_auth_profile_failure(store, profile_id, "unknown", agent_dir=agent_dir)


async def clear_auth_profile_cooldown(
    store: AuthProfileStore,
    profile_id: str,
    agent_dir: str | None = None,
) -> None:
    """Clear cooldown for a profile (manual reset).

    Args:
        store: AuthProfileStore
        profile_id: Profile ID
        agent_dir: Optional agent directory
    """

    def updater(fresh_store: AuthProfileStore) -> bool:
        if not fresh_store.usage_stats or profile_id not in fresh_store.usage_stats:
            return False

        stats = fresh_store.usage_stats[profile_id]
        fresh_store.usage_stats[profile_id] = ProfileUsageStats(
            last_used=stats.last_used,
            error_count=0,
            cooldown_until=None,
            disabled_until=stats.disabled_until,
            disabled_reason=stats.disabled_reason,
            failure_counts=stats.failure_counts,
            last_failure_at=stats.last_failure_at,
        )
        return True

    updated = await update_auth_profile_store_with_lock(
        {"agent_dir": agent_dir, "updater": updater}
    )

    if updated:
        store.usage_stats = updated.usage_stats
        return

    # Fallback: direct update
    if not store.usage_stats or profile_id not in store.usage_stats:
        return

    stats = store.usage_stats[profile_id]
    store.usage_stats[profile_id] = ProfileUsageStats(
        last_used=stats.last_used,
        error_count=0,
        cooldown_until=None,
        disabled_until=stats.disabled_until,
        disabled_reason=stats.disabled_reason,
        failure_counts=stats.failure_counts,
        last_failure_at=stats.last_failure_at,
    )
    save_auth_profile_store(store, agent_dir)


def resolve_profile_unusable_until_for_display(
    store: AuthProfileStore,
    profile_id: str,
) -> int | None:
    """Resolve profile unusable until timestamp for display purposes.

    Args:
        store: AuthProfileStore
        profile_id: Profile ID

    Returns:
        Timestamp (ms) or None
    """
    if not store.usage_stats:
        return None

    stats = store.usage_stats.get(profile_id)
    if not stats:
        return None

    return _resolve_profile_unusable_until(stats)
